feat: add support for nested attribute profile mapping

devtekve opened the pull request:
## Summary

This MR introduces support for nested properties in the OAuth connector for Logto, allowing users to map and retrieve nested attributes from social vendor profiles. This enhancement builds upon the existing functionality, enabling more flexible and comprehensive user profile mappings.

### Key Changes:
- Updated the `userProfileMapping` function to handle nested properties.
- Modified the `ProfileMap` type to support nested attribute paths.
- Added unit tests to verify the correct mapping of nested properties.
- Updated the README to include documentation and examples for nested attribute mappings.

## Why? 
Imagine the social vendor user info returns something like 
```json
{
  "id": "123456",
  "contact": {
    "email": "octcat@github.com",
    "phone": "123-456-7890"
  },
  "details": {
    "name": "Oct Cat",
    "avatar": {
      "url": "avatar.png"
    }
  }
}
```

now we can map it without the need of creating a new connector like this 
```json
{
  "id": "id",
  "name": "details.name",
  "avatar": "details.avatar.url",
  "email": "contact.email",
  "phone": "contact.phone"
}
```

for the final result of 
```json
{
  "id": "123456",
  "name": "Oct Cat",
  "avatar": "avatar.png",
  "email": "octcat@github.com",
  "phone": "123-456-7890"
}
```

## Testing

- Created unit tests to ensure the connector correctly handles nested attributes from social vendor profiles.
- Ran existing unit tests to validate that the existing behavior remains as expected.
- Manually tested the connector by authenticating with a real social vendor account, verifying that nested user
 details are retrieved correctly.

### For my testing I connected with Patreon with the following config

- Authorization Endpoint: `https://www.patreon.com/oauth2/authorize`
- Token Endpoint: `https://www.patreon.com/api/oauth2/token`
- User Info Endpoint: `https://www.patreon.com/api/oauth2/api/current_user`
- Token Endpoint Auth Method: `client_secret_basic`
- Token Endpoint Response Type: `json`
- Scope: `identity identity[email]`
And finally, the Profile Map: 
```json
{
  "id": "data.id",
  "name": "data.attributes.full_name",
  "email": "data.attributes.email",
  "avatar": "data.attributes.image_url"
}
```

<img width="1021" alt="image" src="https://github.com/user-attachments/assets/fe58006b-4f38-40ba-875b-f2ce22c92c28">


## Checklist

- [x] `.changeset` - Added a changeset file to record the new feature introduction.
- [x] unit tests - Developed unit tests for nested property mapping.
- [ ] necessary TSDoc comments - Added TSDoc comments to the new methods and classes introduced for documentation.
- [ ] integration tests - Ran integration tests to ensure proper interaction with nested attributes.


### Footnote
This resolves https://github.com/logto-io/logto/issues/5580

darcyYe commented on packages/connectors/connector-oauth2/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@logto/connector-oauth",
-  "version": "1.4.0",
+  "version": "1.5.0",
```
```suggestion
  "version": "1.4.0",
```
There is no need to manually bump the connector version since we are using "[changeset](https://github.com/changesets/changesets)" to do the version control.

darcyYe reviewed (approved):
LGTM, thanks for your contribution

xiaoyijun commented on packages/connectors/connector-oauth2/src/utils.ts:
```diff
@@ -36,6 +36,19 @@ const accessTokenResponseHandler = async (
 };
 
+// eslint-disable-next-line @typescript-eslint/ban-types
+const getValue = (object: Record<string, unknown> | object, path: string) => {
+  return path.split('.').reduce<unknown>((accumulator, part: string) => {
+    // Check if the accumulator is an object and has the property
+    if (accumulator && typeof accumulator === 'object' && part in accumulator) {
+      // eslint-disable-next-line no-restricted-syntax
+      return (accumulator as Record<string, unknown>)[part];
+    }
+
+    return null;
+  }, object);
+};
```
Maybe we can use something like [_.get](https://underscorejs.org/#get)?
cc @charIeszhao

xiaoyijun replied:
Oh, we have a `get` util method in the `silverhand-io/essentials` package, we can use it directly.
https://github.com/silverhand-io/essentials/blob/master/src/utilities/get.ts

xiaoyijun replied:
Yes, you can directly use the existing method.

devtekve replied:
Understood!

devtekve commented:
Changes made, we are now using the essentials and things work as expected. I did have to tweak the last test because now it throws an error if the key is not found rather than silently make it "undefined". (I prefer the error now that i think of it, so it's much better)

darcyYe commented:
Hi @devtekve , I made changes and pushed a commit to your MR. If we use `get()` method in the code, with the existing profile mapping setup, values that can not be found by profile mapping setup could lead to errors, which is a breaking change. Hence I changed the `get()` method to `getSafe()` method to silently mark the unfound value as `undefined`.

devtekve commented:
> Hi @devtekve , I made changes and pushed a commit to your MR. If we use `get()` method in the code, with the existing profile mapping setup, values that can not be found by profile mapping setup could lead to errors, which is a breaking change. Hence I changed the `get()` method to `getSafe()` method to silently mark the unfound value as `undefined`.



> Hi @devtekve , I made changes and pushed a commit to your MR. If we use `get()` method in the code, with the existing profile mapping setup, values that can not be found by profile mapping setup could lead to errors, which is a breaking change. Hence I changed the `get()` method to `getSafe()` method to silently mark the unfound value as `undefined`.

Makes total sense!. I was under the wrong impression it used to fail as well if the key was not found before the nested attributes. But if that wasn't the case then yeah, good catch!
